crates/compression/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+#![deny(clippy::arithmetic_side_effects)]
+#![deny(clippy::cast_possible_truncation)]
+#![deny(unused_crate_dependencies)]
+#![deny(warnings)]
+
+pub mod compress;
+pub mod decompress;
+mod eviction_policy;
+pub mod ports;
+mod registry;
+
+pub use registry::RegistryKeyspace;
+
+use serde::{
+    Deserialize,
+    Serialize,
+};
+
+use fuel_core_types::{
+    blockchain::header::PartialBlockHeader,
+    fuel_tx::CompressedTransaction,
+    fuel_types::Bytes32,
+};
+use registry::RegistrationsPerTable;
+
+/// Compressed block, without the preceding version byte.
+#[derive(Clone, Serialize, Deserialize)]
+struct CompressedBlockPayloadV0 {
+    /// Temporal registry insertions
+    registrations: RegistrationsPerTable,
+    /// Merkle root of the temporal registry state
+    registrations_root: Bytes32,
+    /// Compressed block header
+    header: PartialBlockHeader,
+    /// Compressed transactions
+    transactions: Vec<CompressedTransaction>,
+}
+
+/// Versioned compressed block.
+#[derive(Clone, Serialize, Deserialize)]
+enum CompressedBlock {
+    V0(CompressedBlockPayloadV0),
+}
+
+#[cfg(test)]
+mod tests {
+    use fuel_core_compression as _;
+    use fuel_core_types::{
+        blockchain::{
+            header::{
+                ApplicationHeader,
+                ConsensusHeader,
+            },
+            primitives::Empty,
+        },
+        fuel_compression::RegistryKey,
+        tai64::Tai64,
+    };
+    use proptest::prelude::*;
+
+    use super::*;
+
+    fn keyspace() -> impl Strategy<Value = RegistryKeyspace> {
+        prop_oneof![
+            Just(RegistryKeyspace::Address),
+            Just(RegistryKeyspace::AssetId),
+            Just(RegistryKeyspace::ContractId),
+            Just(RegistryKeyspace::ScriptCode),
+            Just(RegistryKeyspace::PredicateCode),
+        ]
+    }
+
+    proptest! {
+        /// Serialization for compressed transactions is already tested in fuel-vm,
+        /// but the rest of the block de/serialization is be tested here.
```

Comment:
While it is true, it doesn't change the fact that we need to test compression and decompression here. In the `fuel-tx,`, we test those traits for compression, and their deriving works correctly. Here, we need to test that the logic of the compress and decompress contexts is correct. We need to be sure that we collect the right addresses, IDs, and codes and store them in the database. That we assign correct register keys, and can handle cases when we use the same register key during one compression.

Reply:
Yep. The testing is clearly not sufficient. I've taken this PR back to draft stage until I manage to write those.

Reply:
We can merge this PR if we do not enable block compression in the `fuel-core` to simplify the review process; what do you think?

Reply:
We can. However, I don't trust the code to work correctly, and I'd like the code on `master` to be in a good state. That said, easing up the review process could be worth it.